Compiler error when building python bindings with tracing (mismatched exception specification)

I think this is not hitting out GCC builds (for releases) because clang is being stricter.

The issue is that the Python bindings build with exceptions enabled (they have to) and if they include tracing with `IREE_TRACING_FEATURE_ALLOCATION_TRACKING` (default), then the definition of new must be declared to throw. Just changing in situ causes unbalanced allocations in Tracy, indicating an issue.

```
In file included from /home/stella/src/iree/bindings/python/iree/runtime/vm.cc:7:
In file included from /home/stella/src/iree/bindings/python/iree/runtime/vm.h:15:
In file included from /home/stella/src/iree/iree/vm/api.h:13:
In file included from /home/stella/src/iree/iree/vm/context.h:15:
In file included from /home/stella/src/iree/iree/vm/stack.h:17:
/home/stella/src/iree/iree/base/tracing.h:471:7: error: exception specification in declaration does not match previous declaration
void* operator new(size_t count) noexcept;
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/new:126:26: note: previous declaration is here
_GLIBCXX_NODISCARD void* operator new(std::size_t) _GLIBCXX_THROW (std::bad_alloc)
```

--- a/bindings/python/iree/runtime/CMakeLists.txt
+++ b/bindings/python/iree/runtime/CMakeLists.txt
@@ -34,6 +34,10 @@ iree_pyext_module(
     "vm.cc"
   UNIX_LINKER_SCRIPT
     "unix_version.lds"
+  DEFINES
+    # Pybind code seems to be incompatible with C++ allocation tracing
+    # hooks so disable it.
+    IREE_TRACING_HOOK_CPP_NEW_DELETE=0
   DEPS
     iree::base
     iree::base::cc

--- a/build_tools/cmake/iree_python.cmake
+++ b/build_tools/cmake/iree_python.cmake
@@ -142,11 +142,14 @@ endfunction()
 # MODULE_NAME: Base-name of the module.
 # SRCS: List of source files for the library
 # DEPS: List of other targets the test python libraries require
+# COPTS: List of private compile options
+# DEFINES: List of public defines
+# INCLUDES: Include directories to add to dependencies
 function(iree_pyext_module)
   cmake_parse_arguments(ARG
     ""
     "NAME;MODULE_NAME;UNIX_LINKER_SCRIPT"
-    "SRCS;DEPS;COPTS;INCLUDES"
+    "SRCS;DEPS;COPTS;DEFINES;INCLUDES"
     ${ARGN})
 
   iree_package_ns(_PACKAGE_NS)
@@ -214,6 +217,11 @@ function(iree_pyext_module)
     ${ARG_COPTS}
     ${IREE_DEFAULT_COPTS}
     ${_RTTI_AND_EXCEPTION_COPTS}
+  )
+
+  target_compile_definitions(
+    ${_NAME} PUBLIC
+    ${ARG_DEFINES}
   )
 
   # Link flags.

--- a/iree/base/tracing.cc
+++ b/iree/base/tracing.cc
@@ -203,21 +203,3 @@ void iree_tracing_mutex_after_unlock(uint32_t lock_id) {
 #ifdef __cplusplus
 }  // extern "C"
 #endif  // __cplusplus
-
-#if defined(__cplusplus) &&                                               \
-    (IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_ALLOCATION_TRACKING) && \
-    !IREE_SANITIZER_ADDRESS && !IREE_SANITIZER_MEMORY &&                  \
-    !IREE_SANITIZER_THREAD
-
-void* operator new(size_t count) noexcept {
-  auto ptr = malloc(count);
-  IREE_TRACE_ALLOC(ptr, count);
-  return ptr;
-}
-
-void operator delete(void* ptr) noexcept {
-  IREE_TRACE_FREE(ptr);
-  free(ptr);
-}
-
-#endif  // __cplusplus && IREE_TRACING_FEATURE_ALLOCATION_TRACKING

--- a/iree/base/tracing.h
+++ b/iree/base/tracing.h
@@ -29,6 +29,10 @@
 #include <stdint.h>
 #include <stdlib.h>
 
+#if defined(__cplusplus)
+#include <new>
+#endif
+
 #include "iree/base/attributes.h"
 #include "iree/base/config.h"
 
@@ -177,6 +181,16 @@
 #if IREE_TRACING_MAX_CALLSTACK_DEPTH == 0
 #undef TRACY_CALLSTACK
 #endif  // IREE_TRACING_MAX_CALLSTACK_DEPTH
+
+// By default, hook the C++ new and delete operators if allocation tracing
+// is enabled. We allow this to be controlled indendently because not all
+// C++ code is well behaved with respect to allocations, and Tracy gets
+// very angry when unbalanced. Such code can:
+//   #define IREE_TRACING_HOOK_CPP_NEW_DELETE 0
+// before including this file.
+#if !defined(IREE_TRACING_HOOK_CPP_NEW_DELETE)
+#define IREE_TRACING_HOOK_CPP_NEW_DELETE 1
+#endif
 
 //===----------------------------------------------------------------------===//
 // C API used for Tracy control
@@ -466,10 +480,26 @@ enum {
 #define IREE_TRACE_FREE_NAMED(name, ptr)
 #endif  // IREE_TRACING_FEATURE_ALLOCATION_TRACKING
 
-#if defined(__cplusplus) && \
-    (IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_ALLOCATION_TRACKING)
-void* operator new(size_t count) noexcept;
-void operator delete(void* ptr) noexcept;
+#if defined(__cplusplus) && IREE_TRACING_HOOK_CPP_NEW_DELETE &&           \
+    (IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_ALLOCATION_TRACKING) && \
+    (!IREE_SANITIZER_ADDRESS && !IREE_SANITIZER_MEMORY &&                 \
+     !IREE_SANITIZER_THREAD)
+inline void* operator new(size_t count, const std::nothrow_t&) noexcept {
+  auto ptr = malloc(count);
+  IREE_TRACE_ALLOC(ptr, count);
+  return ptr;
+}
+
+inline void* operator new(size_t count) throw(std::bad_alloc) {
+  auto ptr = malloc(count);
+  IREE_TRACE_ALLOC(ptr, count);
+  return ptr;
+}
+
+inline void operator delete(void* ptr) noexcept {
+  IREE_TRACE_FREE(ptr);
+  free(ptr);
+}
 #endif  // __cplusplus && IREE_TRACING_FEATURE_ALLOCATION_TRACKING
 
 //===----------------------------------------------------------------------===//

